src/core/websocket/ws.ts

```diff
@@ -346,10 +348,44 @@ export class ReconnectingWebSocket {
         throw Error("Invalid URL");
     }
 
+    /**
+     * Handles the final "tear down" process when connection attempts must permanently stop
+     * (e.g., max retries reached, fatal configuration error).
+     * Releases locks, disables reconnection, dispatches final error and close events.
+     *
+     * @param error - The specific Error instance indicating the reason for termination.
+     * @param closeReason - The reason string to use for the final CloseEvent.
+     */
+    private _terminateConnectionAttempt(error: Error, closeReason: string): void {
```

Comment:
naming nit:

`terminateConnectionAttempt` sounds like it is ending a single attempt to connect -- I would pick a name like

`shutdown` or `disconnectForGood` or `closeFinal`